test/Microsoft.AspNetCore.Mvc.DataAnnotations.Test/Internal/DataAnnotationsMetadataProviderTest.cs

```diff
@@ -152,21 +154,55 @@ public void CreateBindingMetadata_EditableAttributeTrue_SetsReadOnlyFalse()
             Assert.False(context.BindingMetadata.IsReadOnly);
         }
 
+        // This is IMPORTANT. Product code needs to use GetName() instead of .Name. It's easy to regress.
+        [Fact]
+        public void CreateDisplayMetadata_DisplayAttribute_NameFromResources_NullLocalizer()
+        {
+            // Arrange
+            var provider = new DataAnnotationsMetadataProvider(stringLocalizerFactory: null);
+
+            var display = new DisplayAttribute()
+            {
+#if USE_REAL_RESOURCES
+                Name = nameof(Test.Resources.DisplayAttribute_Name),
+                ResourceType = typeof(Test.Resources),
+#else
+                Name = nameof(Test.Resources.DisplayAttribute_Name),
+                ResourceType = typeof(TestResources),
+#endif
+            };
+
+            var attributes = new Attribute[] { display };
+            var key = ModelMetadataIdentity.ForType(typeof(string));
+            var context = new DisplayMetadataProviderContext(key, new ModelAttributes(attributes));
+
+            // Act
+            provider.CreateDisplayMetadata(context);
+
+            // Assert
+            Assert.Equal("name from resources", context.DisplayMetadata.DisplayName());
+        }
 
         // This is IMPORTANT. Product code needs to use GetName() instead of .Name. It's easy to regress.
         [Fact]
-        public void CreateDisplayMetadata_DisplayAttribute_NameFromResources()
+        public void CreateDisplayMetadata_DisplayAttribute_NameFromResources_WithLocalizer()
         {
             // Arrange
-            var provider = new DataAnnotationsMetadataProvider();
+            // Nothing on stringLocalizer should be called
+            var stringLocalizer = new Mock<IStringLocalizer>(MockBehavior.Strict);
+            var stringLocalizerFactory = new Mock<IStringLocalizerFactory>();
+            stringLocalizerFactory
+                .Setup(s => s.Create(It.IsAny<Type>()))
+                .Returns(() => stringLocalizer.Object);
+            var provider = new DataAnnotationsMetadataProvider(stringLocalizerFactory.Object);
 
             var display = new DisplayAttribute()
             {
 #if USE_REAL_RESOURCES
                 Name = nameof(Test.Resources.DisplayAttribute_Name),
                 ResourceType = typeof(Test.Resources),
 #else
-                Name = nameof(DataAnnotations.Test.Resources.DisplayAttribute_Name),
+                Name = nameof(Test.Resources.DisplayAttribute_Name),
```

Comment:
This is not a fix at all; change just makes the two sides of the `#if` identical. That should break this test.

When `USE_REAL_RESOURCES` is defined, the test fails because `[Display]` won't use the `internal` property in `Test.Resources`. If the tests now succeed when using desktop or .NET Core, get rid of `TestResources`, `USE_REAL_RESOURCES` mentions, and all use of `TestResources`. But that's completely separate from changes in this PR. Don't mix it in and, for now, stick with the pattern of the existing resource-using tests.